### PR TITLE
Cache internal objects used by the formatting engine.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Drops allocations by 2.2%.  Review a commit at a time.

image

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Engine/TriviaData.cs [hunk 1]
```diff
@@ -18,16 +18,12 @@ internal abstract class TriviaData
...
-    protected TriviaData(SyntaxFormattingOptions options, string language)
+    protected TriviaData(LineFormattingOptions options)
     {
         Options = options;
-        _language = language;
     }
 
-    protected SyntaxFormattingOptions Options { get; }
-    protected string Language => _language;
+    protected LineFormattingOptions Options { get; }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Engine/TriviaData.cs:26]: all consumers only need the limited amount of info exposed by LineFormattingOPtions,  not the full info in SyntaxFormattingOptions.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Engine/AbstractTriviaDataFactory.cs [hunk 2]
```diff
@@ -14,23 +17,73 @@ internal abstract partial class AbstractTriviaDataFactory
...
     protected readonly TreeData TreeInfo;
-    protected readonly SyntaxFormattingOptions Options;
+    protected readonly LineFormattingOptions Options;
 
     private readonly Whitespace[] _spaces;
-    private readonly Whitespace?[,] _whitespaces = new Whitespace[LineBreakCacheSize, IndentationLevelCacheSize];
+    private readonly Whitespace[,] _whitespaces;
 
-    protected AbstractTriviaDataFactory(TreeData treeInfo, SyntaxFormattingOptions options)
+    protected AbstractTriviaDataFactory(TreeData treeInfo, LineFormattingOptions options)
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Engine/AbstractTriviaDataFactory.cs:29]: I'll check. But it is exposed as a property on this type.

[the remark this answers is not in the document]
Comment (ToddGrun) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Engine/AbstractTriviaDataFactory.cs:29]: Ah, yes, that is used by the derived factories.

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Pull request merged**